### DISCUSSION TITLE
Skip module resolution w/ Metro bundler

### DISCUSSION THREAD

**Discussion by Akronae** [expo/expo#21736] [Archived: Questions]
Using the new Metro Bundler and, in my case, `react-native-maps`, I have compatibilities issues when building for web.

**Summup:**
How can I tell metro/babel to alias `react-native-maps` to something like `null-loader` on web build only?

**Details:**

I have some code requiring `react-native-maps` like so:
```jsx
if (Platform != 'web') {
  const MapView = require(`react-native-maps`).default
  // ...
}
```
When building for native everything works fine, but during a web build even though this code is not executed at runtime it seems like Metro is still trying to resolve `react-native-maps` and by doing so the compilation is broken, I'll have errors like these:
```
Unable to resolve "../Utilities/Platform" from "node_modules/react-native/Libraries/ReactPrivate/ReactNativePrivateInterface.js"
```
(because `Platform.web.js` does not exists)

However if I require `react-native-maps` "obfucasting" the path for Metro not to track the module like so:
```ts
export function requireNativeOnly<T = any>(path: string): T {
  return ({
    'require': require,
  })['require'](path)
}
// ...
if (Platform != 'web') {
  const MapView = requireNativeOnly(`react-native-maps`).default
  // ...
}
```
My web build will work perfectly fine but on native at runtime when this module is required (through `requireNativeOnly`) I'll have errors saying this module cannot be found:
```
Unknown named module: "react-native-maps"
```

How can I tell expo/babel to alias `react-native-maps` to something like `null-load` on web build only?

**Comment by adamhari** [comment]
I'm wondering the same thing. I've tried three different methods of [configuring metro's resolver](https://facebook.github.io/metro/docs/resolution/) to alias like webpack would and I haven't had any success:

```
const config = getDefaultConfig(__dirname);

const modulesToAlias = {
  '@sentry/react-native': '@sentry/browser',
  'lottie-react-native': 'react-native-web-lottie',
  'victory-native': 'victory',
};

module.exports = {
  ...config,
  resolver: {
  ...config.resolver,
    // 1
    extraNodeModules: modulesToAlias,
    // 2
    extraNodeModules: Object.entries(modulesToAlias).reduce((acc, [k, v]) => {
      acc[k] = path.join(__dirname, 'node_modules', v);
      return acc;
    }, {}),
    // 3
    resolveRequest: (context, moduleName, platform) => {
      console.log({ moduleName, platform });

      if (platform === 'web') {
        Object.entries(modulesToAlias).forEach(([k, v]) => {
          if (k === moduleName) {
            return {
              filePath: path.join(__dirname, 'node_modules', v),
              type: 'sourceFile',
            };
          }
        });
      }

      return context.resolveRequest(context, moduleName, platform);
    },
  },
};
```

**Reply by alexandresilvacarreira** [reply]
**EDIT:** I'm dumb, was pointing to a directory instead of a file inside that directory. The `../Utilities/Platform` module should be aliased with  `react-native-web/dist/exports/Platform/index.js`. But now there are of course a lot more modules that need to be aliased. It seems that it's not as simple as pointing all imports web-related to `react-native-web`...

I am facing a similar problem and I'm at my wits end. I believe I was able to configure Metro to properly use the alias I want, using `mergeConfig` to merge the default config with my custom config object as per the [Metro docs](https://metrobundler.dev/docs/configuration/#merging-configurations). However, now I get an error which I think is in related to Metro not being able to access the aliased directory, which is weird since it's configured to watch all folders in the project. 

Regarding the `../Utilities/Platform` module, there are more [people falling into this rabbit hole](https://forums.expo.dev/t/expo-web-module-not-found-cant-resolve-utilities-platform/53797/4) and the solution seems to be to point to the proper library under `react-native-web` with an alias, thus making it look under `react-native-web/dist/exports/Platform/index,js` instead, which is where the `Platform` library resides in `react-native-web.`

Here's my `metro.config.js`:

```
const { getDefaultConfig } = require('expo/metro-config');
const { mergeConfig } = require('metro-config');
const path = require('path');

/** @type {import('expo/metro-config').MetroConfig} */
const defaultConfig = getDefaultConfig(__dirname, {
  // [Web-only]: Enables CSS support in Metro.
  isCSSEnabled: true,
});

// Custom resolver function
const customResolver = function (context, moduleName, platform) {
  if (platform === 'web' && moduleName === '../Utilities/Platform') {
    const relativePath = path.resolve(
      'node_modules/react-native-web/dist/exports/Platform/index,js'
    );
    return {
      type: 'sourceFile',
      filePath: relativePath,
    };
  }
  // Default behavior for other module resolutions
  return context.resolveRequest(context, moduleName, platform);
};

defaultConfig.resolver.sourceExts.push('mjs');

const config = mergeConfig(defaultConfig, {
  watchFolders: [path.resolve(__dirname)],
  resolver: {
    resolveRequest: customResolver,
    useWatchman: false,
    unstable_enablePackageExports: true,
    unstable_enableSymlinks: true,
    extraNodeModules: {
      'react-native': process.env.WEB ? 'react-native-web' : 'react-native',
    },
    nodeModulesPaths: [path.resolve(__dirname, 'node_modules')],
  },
});

module.exports = config;
```
_Edit: this error was fixed by pointing to the actual file instead of the directory, see my edit at the top_

The error I'm getting now, when running `npx expo start -c --web` is the following:

```
error: ReferenceError: SHA-1 for file C:\projects\RnWebApp\node_modules\react-native-web\dist\exports\Platform (C:\projects\RnWebApp\node_modules\react-native-web\dist\exports\Platform) is not computed.
         Potential causes:
           1) You have symlinks in your project - watchman does not follow symlinks.
           2) Check `blockList` in your metro.config.js and make sure it isn't excluding the file path.
    at DependencyGraph.getSha1 (C:\projects\RnWebApp\node_modules\metro\src\node-haste\DependencyGraph.js:217:13)
    at Transformer._getSha1 (C:\projects\RnWebApp\node_modules\metro\src\Bundler.js:26:26)
    at Transformer.transformFile (C:\projects\RnWebApp\node_modules\metro\src\DeltaBundler\Transformer.js:106:19)
    at Bundler.transformFile (C:\projects\RnWebApp\node_modules\metro\src\Bundler.js:60:30)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.transform (C:\projects\RnWebApp\node_modules\metro\src\lib\transformHelpers.js:143:12)
    at async Graph._processModule (C:\projects\RnWebApp\node_modules\metro\src\DeltaBundler\Graph.js:257:20)
    at async Graph._addDependency (C:\projects\RnWebApp\node_modules\metro\src\DeltaBundler\Graph.js:372:20)
    at async Promise.all (index 4)
    at async Graph._processModule (C:\projects\RnWebApp\node_modules\metro\src\DeltaBundler\Graph.js:322:5)
```

I have enabled the `unstable_enableSymlinks` as well as `unstable_enablePackageExports` and disabled watchman, so I don't know what else to try. Webpack is deprecated, so I'd like to avoid using it. I'm on Expo SDK 49.0.15 and all the installed web dependencies are compatible with this version.

_Edit: formatting, grammar_

**Reply by brunolucena** [reply]
Thanks for the solution!
I've used your solution but ignored the entire package causing the problem, which in my case were react-native-pager-view and react-native-text-input-mask.
This is my metro.config.js with the solution

```
const { getDefaultConfig } = require("expo/metro-config");
const OriginalResolver = require("metro-resolver");
const path = require("path");

const webBlacklistedModules = [
  "react-native-pager-view",
  "react-native-text-input-mask",
];

/** @type {import('expo/metro-config').MetroConfig} */
const config = getDefaultConfig(__dirname);

config.resolver.resolveRequest = (context, moduleName, platform) => {
  if (platform === "web" && webBlacklistedModules.includes(moduleName)) {
    return {
      type: "empty",
    };
  }
  // Default behavior for other module resolutions
  return context.resolveRequest(context, moduleName, platform);
};

module.exports = config;
```

**Reply by ccdwyer** [reply]
@brunolucena I owe you and @alexandresilvacarreira a beer. Awesome solution to the problem I was hitting.
